gobblin-core/src/main/java/gobblin/publisher/BaseDataPublisher.java

```diff
@@ -299,14 +270,75 @@ private ParallelRunner getParallelRunner(FileSystem fs) {
     return this.parallelRunners.get(uri);
   }
 
-  @Override
-  public void publishMetadata(Collection<? extends WorkUnitState> states) throws IOException {
-    // Nothing to do
+  private WorkUnitPublishGroup getWorkUnitPublishGroup(WorkUnitState state) throws IOException {
+    for (int branchId = 0; branchId < this.numBranches; branchId++) {
+      // The directory where the workUnitState wrote its output data.
+      Path writerOutputDir = WriterUtils.getWriterOutputDir(state, this.numBranches, branchId);
+
+      if (!this.writerFileSystemByBranches.get(branchId).exists(writerOutputDir)) {
+        LOG.warn(String.format("Branch %d of WorkUnit %s produced no data", branchId, state.getId()));
+        continue;
+      }
+
+      // The directory where the final output directory for this job will be placed.
+      // It is a combination of DATA_PUBLISHER_FINAL_DIR and WRITER_FILE_PATH.
+      Path publisherOutputDir = getPublisherOutputDir(state, branchId);
+
+      return new WorkUnitPublishGroup(branchId, writerOutputDir, publisherOutputDir);
```

Comment:
Looking at this a little more, it seems as if the each workunit can be part of multiple branches and it should return a set of groups, one per branch the workunit is part of.